scripts/document_types/gene_annotator.py

```diff
@@ -0,0 +1,387 @@
+
+        if self.__verbose : print('[Info] Input data validation passed.')
+
+    def __get_Ensembl_gene_annotation(self):
+
+        # list with all the genes and the available annotation:
+        EnsemblAnnotationContainer = []
+
+        def parse_annotation(row):
+            fields = row.split("\t")
+            
+            if len(fields) < 8: 
+                print(row)
+                return()
+
+            # Extracting coordinates:
+            geneAnnot = {
+                'seq_region_name' : fields[0],
+                'start' : fields[3],
+                'end' : fields[4],
+            }
+            
+            # Parse description column of the gff file:
+            for annot in fields[8].split(";"):
+                key, value = annot.split("=")
+                if key == 'ID':
+                    geneAnnot['id'] = value.split(":")[1]
+                if key == 'Name':
+                    geneAnnot['display_name'] = value
+                if key == 'biotype':
+                    geneAnnot['biotype'] = value
+                if key == 'description':
+                    geneAnnot['description'] = urllib.unquote(value.split(" [")[0])
+                else:
+                    continue
+            
+            # Adding annotation to the container
+            return(geneAnnot)
+
+        # Using bash to filter the gff3 file with 2.8M lines
+        filteredLines = subprocess.Popen(['zgrep', 'ID=gene', self.__EnsemblFile],stdout=subprocess.PIPE)
+
+        # Processing output:
+        for line in iter(filteredLines.stdout.readline,''):
+            EnsemblAnnotationContainer.append(parse_annotation(line.rstrip()))
+   
+        # Pooling annotations into a pandas dataframe and format table:
+        EnsAnnotDf = pd.DataFrame.from_records(EnsemblAnnotationContainer)
+
+        # Setting start and end columns as integers:
+        EnsAnnotDf.start = EnsAnnotDf.start.astype(int)
+        EnsAnnotDf.end = EnsAnnotDf.end.astype(int)
+
+        # Filtering df to only a selected list of columns:
+        EnsAnnotDf = EnsAnnotDf[['seq_region_name','start', 'end', 'id', 'display_name', 'biotype', 'description']].sort_values(['seq_region_name', 'start'])
+        
+        # Reindex:
+        EnsAnnotDf.index = EnsAnnotDf.id.tolist()
+        
+        if self.__verbose: print("[Info] Ensembl annotations are extracted for %s genes" % len(EnsAnnotDf))
+        
+        self.__ensembl_data = EnsAnnotDf
+
+    def __get_HGNC(self):
+        '''
+        Extract HGNC data from /nfs. Extracting relevant columns and formats
+
+        1. Select certain fields: alternative names/synonyms + cros ref IDs
+        2. Pool these names together + ensembl ID and entrez ID are separate columns.
+        3. Adding dataframe to object.
+        '''
+
+        # The file name needs to be read from the environment:
+        HGNC_file = '/nfs/ftp/pub/databases/genenames/new/tsv/non_alt_loci_set.txt'
```

Comment:
This filepath reference should be removed.

Reply:
It's removed.